File: TP/TP-C/TP2/sub.c
#include "zhelpers.h"

int main (int argc, char *argv [])
{
    //  Socket to talk to server
    printf ("Collecting updates from weather servers...\n");
    void *context = zmq_ctx_new ();
    void *subscriber_us = zmq_socket (context, ZMQ_SUB);
    void *subscriber_pt = zmq_socket (context, ZMQ_SUB);
    
    int rc = zmq_connect (subscriber_us, "tcp://localhost:5556");
    assert (rc == 0);
    rc = zmq_connect (subscriber_pt, "tcp://localhost:5557");
    assert (rc == 0);

    const char *filter_us = "10001 ";
    const char *filter_pt = "4200 ";
    rc = zmq_setsockopt (subscriber_us, ZMQ_SUBSCRIBE, filter_us, strlen (filter_us));
    assert (rc == 0);
    rc = zmq_setsockopt (subscriber_pt, ZMQ_SUBSCRIBE, filter_pt, strlen (filter_pt));
    assert (rc == 0);

    zmq_pollitem_t items [] = {
        { subscriber_us, 0, ZMQ_POLLIN, 0 },
        { subscriber_pt, 0, ZMQ_POLLIN, 0 }
    };

    //  Process 100 updates
    int update_nbr;
    long total_temp[2] = {0};
    int count[2] = {0};

    for (update_nbr = 0; count[0] < 100 || count[1] < 100; update_nbr++) {
        zmq_poll (items, 2, -1);
        int zipcode, temperature, relhumidity;

        if (items[0].revents & ZMQ_POLLIN && count[0] < 100) {
            char *string = s_recv (subscriber_us);
            sscanf (string, "%d %d %d", &zipcode, &temperature, &relhumidity);
            total_temp[0] += temperature;
            count[0]++;
            printf("Received US update: %s\n", string);
            free (string);
        }

        if (items[1].revents & ZMQ_POLLIN && count[1] < 100) {
            char *string = s_recv (subscriber_pt);
            sscanf (string, "%d %d %d", &zipcode, &temperature, &relhumidity);
            total_temp[1] += temperature;
            count[1]++;
            printf("Received PT update: %s\n", string);
            free (string);
        }
    }

    printf("Count US: %d\n", count[0]);
    printf("Count PT: %d\n", count[1]);

    printf ("Average temperature for US zipcode '%s' was %dF\n", filter_us, (int) (total_temp[0] / count[0]));
    printf ("Average temperature for PT zipcode '%s' was %dF\n", filter_pt, (int) (total_temp[1] / count[1]));

    zmq_close (subscriber_us);
    zmq_close (subscriber_pt);
    zmq_ctx_destroy (context);
    return 0;
}

File: TP/TP-Python/TP2/sub.py


#
#   Weather update client
#   Connects SUB socket to tcp://localhost:5556
#   Collects weather updates and finds avg temp in zipcode
#

import sys
import zmq


#  Socket to talk to server
context = zmq.Context()
subscriber_us = context.socket(zmq.SUB)
subscriber_pt = context.socket(zmq.SUB)

print("Collecting updates from weather server...")
subscriber_us.connect("tcp://localhost:5556")
subscriber_pt.connect("tcp://localhost:5557")

# Subscribe to zipcode, default is NYC, 10001
filter_us = "10001 "
filter_pt = "4200 "

subscriber_us.setsockopt_string(zmq.SUBSCRIBE, filter_us)
subscriber_pt.setsockopt_string(zmq.SUBSCRIBE, filter_pt)

# Initialize poll set
poller = zmq.Poller()
poller.register(subscriber_us, zmq.POLLIN)
poller.register(subscriber_pt, zmq.POLLIN)

total_temp_us = 0
total_temp_pt = 0
count_us = 0
count_pt = 0

# Collect 100 updates from each source
while count_us < 100 or count_pt < 100:
    socks = dict(poller.poll())

    # Check if a message is available from the U.S. subscriber
    if subscriber_us in socks and count_us < 100:
        message = subscriber_us.recv_string()
        zipcode, temperature, relhumidity = message.split()
        total_temp_us += int(temperature)
        count_us += 1
        print(f"U.S. Received: {zipcode} {temperature} {relhumidity}")

    # Check if a message is available from the Portugal subscriber
    if subscriber_pt in socks and count_pt < 100:
        message = subscriber_pt.recv_string()
        zipcode, temperature, relhumidity = message.split()
        total_temp_pt += int(temperature)
        count_pt += 1
        print(f"Portugal Received: {zipcode} {temperature} {relhumidity}")

# Calculate and print the average for each location
print(f"Average temperature for U.S. is {total_temp_us / count_us}")
print(f"Average temperature for Portugal is {total_temp_pt / count_pt}")

